File: apps/api/plane/hw/serializers/agent.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

from rest_framework import serializers

from plane.app.serializers import BaseSerializer
from plane.hw.models import AgentProfile, AgentRun, AgentRunActivity, AgentConversation, AgentConversationMessage


class AgentProfileSerializer(BaseSerializer):
    class Meta:
        model = AgentProfile
        fields = [
            "id",
            "user_id",
            "workspace_id",
            "webhook_url",
            "event_triggers",
            "is_active",
            "display_name",
            "description",
            "agent_type",
            "created_at",
            "updated_at",
        ]
        read_only_fields = [
            "id",
            "user_id",
            "workspace_id",
            "agent_type",
            "created_at",
            "updated_at",
        ]


class AgentProfileCreateSerializer(BaseSerializer):
    class Meta:
        model = AgentProfile
        fields = [
            "display_name",
            "description",
            "webhook_url",
            "webhook_secret",
            "event_triggers",
            "agent_type",
        ]


class AgentRunSerializer(BaseSerializer):
    class Meta:
        model = AgentRun
        fields = [
            "id",
            "agent_id",
            "workspace_id",
            "project_id",
            "issue_id",
            "status",
            "stale_timeout",
            "last_activity_at",
            "completed_at",
            "trigger_metadata",
            "conversation_id",
            "created_at",
            "updated_at",
        ]
        read_only_fields = [
            "id",
            "agent_id",
            "workspace_id",
            "project_id",
            "issue_id",
            "last_activity_at",
            "completed_at",
            "conversation_id",
            "created_at",
            "updated_at",
        ]


class AgentRunCreateSerializer(BaseSerializer):
    agent_id = serializers.UUIDField(write_only=True)
    project_id = serializers.UUIDField(write_only=True, required=False, allow_null=True)
    issue_id = serializers.UUIDField(write_only=True, required=False, allow_null=True)

    class Meta:
        model = AgentRun
        fields = [
            "agent_id",
            "project_id",
            "issue_id",
            "trigger_metadata",
        ]


class AgentRunActivitySerializer(BaseSerializer):
    class Meta:
        model = AgentRunActivity
        fields = [
            "id",
            "run_id",
            "activity_type",
            "content",
            "metadata",
            "is_ephemeral",
            "created_at",
            "updated_at",
        ]
        read_only_fields = [
            "id",
            "run_id",
            "is_ephemeral",
            "created_at",
            "updated_at",
        ]


class AgentConversationSerializer(BaseSerializer):
    class Meta:
        model = AgentConversation
        fields = [
            "id",
            "workspace_id",
            "user_id",
            "title",
            "is_active",
            "created_at",
            "updated_at",
        ]
        read_only_fields = [
            "id",
            "workspace_id",
            "user_id",
            "created_at",
            "updated_at",
        ]


class AgentConversationMessageSerializer(BaseSerializer):
    class Meta:
        model = AgentConversationMessage
        fields = [
            "id",
            "conversation_id",
            "role",
            "content",
            "run_id",
            "created_at",
            "updated_at",
        ]
        read_only_fields = [
            "id",
            "conversation_id",
            "run_id",
            "created_at",
            "updated_at",
        ]

File: apps/web/hw/types/agent.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

export type TAgentRunStatus = "created" | "in_progress" | "completed" | "failed" | "stopped" | "stale";
export type TAgentActivityType = "thought" | "action" | "response" | "elicitation" | "error";
export type TAgentType = "external" | "builtin";

export type TAgentRun = {
  id: string;
  agent_id: string;
  workspace_id: string;
  project_id: string | null;
  issue_id: string | null;
  status: TAgentRunStatus;
  stale_timeout: number;
  last_activity_at: string;
  completed_at: string | null;
  trigger_metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
};

export type TAgentRunActivity = {
  id: string;
  run_id: string;
  activity_type: TAgentActivityType;
  content: string;
  metadata: Record<string, unknown>;
  is_ephemeral: boolean;
  created_at: string;
  updated_at: string;
};

export type TCreateActivityPayload = {
  activity_type: TAgentActivityType;
  content: string;
  metadata?: Record<string, unknown>;
};

export type TAgentProfile = {
  id: string;
  user_id: string;
  workspace_id: string;
  agent_type: TAgentType;
  webhook_url: string;
  webhook_secret: string;
  event_triggers: Record<string, unknown>;
  is_active: boolean;
  display_name: string;
  description: string;
  created_at: string;
  updated_at: string;
};

export type TCreateAgentProfilePayload = {
  display_name: string;
  description: string;
  webhook_url: string;
  webhook_secret: string;
  event_triggers?: Record<string, unknown>;
};

export type TAgentProfileCreateResponse = TAgentProfile & {
  api_token: string;
};
